pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,471 @@
+		containerState == kubecontainer.ContainerStateRunning {
+		return true
+	}
+
+	if event.ContainerEventType == runtimeapi.ContainerEventType_CONTAINER_STOPPED_EVENT &&
+		containerState == kubecontainer.ContainerStateExited {
+		return true
+	}
+
+	if event.ContainerEventType == runtimeapi.ContainerEventType_CONTAINER_DELETED_EVENT {
+		return true
+	}
+
+	return false
+}
+
+func (e *EventedPLEG) isValidateEvent(event *runtimeapi.ContainerEventResponse, podStatus *kubecontainer.PodStatus) bool {
+	if len(podStatus.ContainerStatuses) == 0 && len(podStatus.SandboxStatuses) == 0 &&
+		event.ContainerEventType == runtimeapi.ContainerEventType_CONTAINER_DELETED_EVENT {
+		return true
+	}
+
+	if e.isSandboxEvent(event, podStatus) {
+		return e.isValideSandboxEvent(event, podStatus)
+	} else {
+		return e.validateContainerEvent(event, podStatus)
+	}
+}
+
+func (e *EventedPLEG) updatePodStatus(event *runtimeapi.ContainerEventResponse) {
+	podID := types.UID(event.PodSandboxMetadata.Uid)
+	podName := event.PodSandboxMetadata.Name
+	podNamespace := event.PodSandboxMetadata.Namespace
+
+	timestamp := e.clock.Now()
+	status, err := e.runtime.GetPodStatus(podID, podName, podNamespace)
+
+	if err != nil {
+		// nolint:logcheck // Not using the result of klog.V inside the
+		// if branch is okay, we just use it to determine whether the
+		// additional "podStatus" key and its value should be added.
+		if klog.V(6).Enabled() {
+			klog.ErrorS(err, "Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName), "podStatus", status)
+		} else {
+			klog.ErrorS(err, "Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName))
+		}
+	} else {
+		if klogV := klog.V(6); klogV.Enabled() {
+			klogV.InfoS("Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName), "podStatus", status)
+		} else {
+			klog.V(4).InfoS("Evented PLEG: Write status", "pod", klog.KRef(podNamespace, podName))
+		}
+		// Preserve the pod IP across cache updates if the new IP is empty.
+		// When a pod is torn down, kubelet may race with PLEG and retrieve
+		// a pod status after network teardown, but the kubernetes API expects
+		// the completed pod's IP to be available after the pod is dead.
+		status.IPs = e.getPodIPs(podID, status)
+	}
+
+	// TODO - Disabling it until we fix the issue with inconsistent pod status
+	// if !e.isValidateEvent(event, status) {
+	// 	klog.ErrorS(err, "Evented PLEG: Invalid event received", "event", event, "podStatus", status)
+	// 	e.Relist() // Force relisting to make sure the pod statuses are up to date.
+	// 	return
+	// }
+
+	e.updateRunningPodMetric(status)
+	e.updateRunningContainerMetric(status)
+
+	if event.ContainerEventType == runtimeapi.ContainerEventType_CONTAINER_DELETED_EVENT {
+		for _, sandbox := range status.SandboxStatuses {
+			if sandbox.Id == event.ContainerId {
+				e.cache.Delete(podID)
+			}
+		}
+	} else {
+		e.cache.Set(podID, status, err, timestamp)
+	}
+	e.cache.UpdateTime(timestamp)
```

Comment:
the cache timestamp was originally designed to illustrate the relist has completed and all the pods have refreshed, now for evented pleg, each pod update refreshes the cache timestamp looks worrying if we still using `GetNewerThan` in podworker